pymarlin/utils/checkpointer/checkpoint_utils.py

```diff
@@ -291,3 +291,99 @@ def check_mk_dir(self, dirpath: str) -> None:
             "is not a directory"
+
+
+@dataclass
+class BestCheckpointerArguments(DefaultCheckpointerArguments):
+    """Additional arguments for checkpointer
+
+    metric_name: name of metric where minimal is defined as best. Must be a registered buffer in module interface
+    save_every_epoch: whether to produce a checkpointer every epoch in addition to latest and best.
+    load_best: whether to load best or latest checkpoint. Default behavior is to load latest.
+    """
+    metric_name: str = "val_perplexity"
+    init_metric_val: Optional[float] = None
+    criteria: Optional[Tuple[str, Callable]] = "min"
+    save_every_epoch: bool = False  # not usually necessary in practice
+    load_best: bool = False  # default to load latest
+
+
+class BestCheckpointer(DefaultCheckpointer):
+    """
+    Saves best and latest checkpoint. Best checkpoint is defined as the smallest value of a given parameter in the
+    module interface. Therefore this checkpointer works by relying on the parameter defined in metric_name existing as a
+    single value. By default it checks "val_perplexity" which is a registered buffer in `AbstractUserMessageReplyModule`
+    that gets updated after every call to `on_end_val_epoch`.
+    """
+    def __init__(self, args: BestCheckpointerArguments):
+        super().__init__(args)
+        self.best_checkpoint_name = f"{self.args.file_prefix}_best_checkpoint.{self.args.file_ext}"
+        self.latest_checkpoint_name = f"{self.args.file_prefix}_latest_checkpoint.{self.args.file_ext}"
+        if self.args.criteria == 'min':
+            self.criteria_func = lambda new, old: new < old
+            self.best_metric = float('inf')
+        elif self.args.criteria == 'max':
+            self.criteria_func = lambda new, old: new > old
+            self.best_metric = -float('inf')
+        else:
+            self.criteria_func = self.args.criteria
+            self.best_metric = self.args.init_metric_value
+
+        if self.args.init_metric_value is not None:
+            self.best_metric = self.args.init_metric_value
+
+    def save(self, checkpoint_state: Checkpoint, index: int, force=False) -> str:
+        """
+        Saves trainer, optimizer, and module interface state.
+
+        Args:
+            checkpoint_state: instance of `Checkpoint` which contains trainer, optimizer, and module interface state
+            index: current epoch number
+            force: whether to force a save even if period of checkpointing does not line up
+
+        Returns:
+            list of paths checkpoint state was saved to
+        """
+        paths = []
+        if self.args.save_every_epoch:
+            paths.append(super().save(checkpoint_state, index, force))
+        if self.args.checkpoint and ((index % self.args.period == 0) or force):
+            # TODO grab this from logged metrics instead
+            metric = float(checkpoint_state.module_interface_state[self.args.metric_name])
```

Comment:
this part feels hacky if not pulled from metrics in terms of the design patterns used here, but either way I think adding logging here would be good (available metrics, metric selected and its value)